inc/xmath.h

```diff
@@ -1449,116 +1471,214 @@ struct quat : public vec<4, QS>
-	                         const XMTYPE  sphere_r)
+}; // namespace intersect
+
+namespace filter
+{
+
+/**
+ * @brief      { struct_description }
+ *
+ * @tparam     Z     { description }
+ * @tparam     X     { description }
+ * @tparam     U     { description }
+ * @tparam     S     { description }
+ */
+template <size_t Z, size_t X, size_t U, typename S = XMTYPE>
+struct kalman
+{
+	struct
 	{
-		auto l  = sphere_o - ray_o;
-		auto s  = l.dot(ray_d);
-		auto l2 = l.dot(l);
-		auto r2 = sphere_r * sphere_r;
+		mat<X, 1, S> state;
+		mat<X, X, S> covariance = mat<X, X, S>::I();
+		mat<Z, 1, S> measurement_residual;
+	} estimated;
 
-		if (s < 0 && l2 > r2) { return NAN; }
+	mat<X, X, S> I_xx = mat<X, X, S>::I();
 
-		auto m2 = l2 - (s * s);
+	kalman(vec<X, S>& state = {}) { estimated.state = {state}; }
 
-		if (m2 > r2) { return NAN; }
+	/**
+	 * @brief      { function_description }
+	 *
+	 * @param[in]  state_transition  The state transition
+	 * @param[in]  control_to_state  The control to state
+	 * @param[in]  control           The control
+	 */
+	void time_update(const mat<X, X, S>& state_transition,
+	                 const mat<X, U, S>& control_to_state,
+	                 const mat<U, 1, S>& control,
+	                 const mat<X, X, S>& process_noise_covariance = {})
+	{
+		// Advance state in time by state_transition matrix. Apply control
+		// vector transformed into state space to state.
+		estimated.state = state_transition * estimated.state;
+		estimated.state += control_to_state * control;
 
-		auto q = sqrt(r2 - m2);
+		const auto& F = state_transition;
+		auto&       P = estimated.covariance;
+
+		// Inflate covariance. TODO: study
+		P = F * P * F.transpose() + process_noise_covariance;
+	}
 
-		auto b = l2 > r2;
-		return b * (s - q) + !b * (s + q);
+	/**
+	 * @brief      { function_description }
+	 *
+	 * @param[in]  state_transition              The state transition
+	 * @param[in]  state_to_measurement          The state to measurement
+	 * @param[in]  measurement_noise_covariance  The measurement noise
+	 * covariance
+	 * @param[in]  measurement                   The measurement
+	 */
+	void measurement_update(
+	    const mat<X, X, S>& state_transition,
+	    const mat<Z, X, S>& state_to_measurement,
+	    const mat<Z, 1, S>& measurement,
+	    const mat<Z, Z, S>& measurement_noise_covariance = {})
+	{
+		const mat<X, X, S>& F     = state_transition;
+		const mat<Z, X, S>& H     = state_to_measurement;
+		const mat<X, Z, S>  H_T   = state_to_measurement.transpose();
+		auto&               P     = estimated.covariance;
+		auto&               x_hat = estimated.state;
+
+		// Compute error between most recent measurement, and current estimated
+		// state transformed into measurement domain
+
+		auto meas_res_pre = measurement - state_to_measurement * x_hat;
+
+		// TODO: study
+		mat<Z, Z, S> covariance_res_pre =
+		    H * P * H_T + measurement_noise_covariance;
+
+		// TODO: study
```

Comment:
```suggestion
		// Compute Kalman gain
```

Reply:
I know it's the kalman gain. But I want to study it deeply 😆